# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Vaibhavrajput326/flyrank.ai/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [6]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    # find the repo root from wherever this kernel started
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found — are you at the repo root?"
print("Starter data found. You're ready.")

Working dir: /content/flyrank-ml-internship-starter/flyrank-ml-internship-starter
Starter data found. You're ready.


## 1. # My Lane and Why

I chose the Content Refresh Prediction lane.

The goal is to identify pages that are likely to decline in performance and should be refreshed first. Many companies have thousands of pages and cannot manually inspect every page. A prediction system can help prioritize which pages need attention.

This problem is valuable because content teams have limited time and resources. Identifying declining pages early can help prevent traffic loss and improve decision-making.My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [7]:
import pandas as pd

df = pd.read_csv(
    "/content/flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv"
)

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])



Rows: 30000
Columns: 44


The Question
Research Question
Can we identify web pages that are likely to decline in performance so that content teams can prioritize updates efficiently?

Decision
Which pages should be refreshed first?

Action
The content team updates, rewrites, or optimizes the selected pages.

Cost of a Wrong Recommendation
False Positive: A healthy page is refreshed unnecessarily, wasting time and resources.

False Negative: A declining page is ignored, leading to traffic and business opportunity loss.

Why ML Can Help
The dataset contains many features describing page performance and behavior. Humans cannot manually analyze thousands of pages efficiently. Machine learning can identify patterns and rank pages according to their risk of decline.## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

print(df["trend_direction"].value_counts())


trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64


Quick Look at the Data
The starter dataset contains 30,000 pages and data from 32 different clients.

Out of 30,000 pages, 16,262 pages are marked as "down", which means approximately 54.21% of the pages show declining trends.

This is important because more than half of the pages in the dataset are experiencing decline. In a real business setting, manually reviewing thousands of pages would be difficult and time-consuming.

These numbers suggest that predicting which pages need content refresh is a meaningful and practical problem where machine learning can help prioritize work and support decision-making.## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
print(
    "Unique clients:",
    df["client_id"].nunique()
)


Unique clients: 32


Careful Words
This project does not claim that refreshing a page will automatically increase traffic.

The goal is only to identify pages that may need attention based on historical patterns.

The model can support human decision-making, but it should not replace expert judgment.## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
declining_pct = (
    (df["trend_direction"] == "down").mean()
    * 100
)

print(
    f"Percentage of declining pages: {declining_pct:.2f}%"
)

Percentage of declining pages: 54.21%


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.